SourceX/controls/game_controls.cpp

```diff
@@ -117,6 +117,9 @@ bool GetGameAction(const SDL_Event &event, GameAction *action)
 	if (CurrentGameController() != nullptr && event.type >= SDL_JOYAXISMOTION && event.type <= SDL_JOYBUTTONUP) {
 		return true;
 	}
+	if (event.type == SDL_CONTROLLERAXISMOTION) {

```

Comment:
Why is this necessary?

This event is mapped to `ControllerButton::AXIS_TRIGGERLEFT/RIGHT`.

Mapped here:
https://github.com/glebm/devilutionX/blob/1eecc4c7b8a2ee67cfb0f9b92b94f82cc5b3a6c8/SourceX/controls/devices/game_controller.cpp#L15-L24

And here:
https://github.com/glebm/devilutionX/blob/1eecc4c7b8a2ee67cfb0f9b92b94f82cc5b3a6c8/SourceX/controls/controller.cpp#L14-L17

Looks like you removed the second check but why?

Reply:
It didn't work and was causing events to be triggered incorrectly. Since it is an axis pressing it will cause a series of events with `value` range from 0~32767, and not necessarily in a fully linear fashion. 0 is the last event and will happen when the button is fully released, but mapping 0 to `up`, means that all other events are seen as down, even the once that are fired as you release the button. Instead, I have mapped it so that pressing it past 50% is seen as down, and releasing it is seen as up, and only allowing it to transition between states if the other state has been set first.

https://wiki.libsdl.org/SDL_ControllerAxisEvent

The reason for the specific line is because the intermediate movement isn't directly mapped to a button any longer, but I now hide the cursor when a game controller event is detected, and show it when a non-game controller event is seen, so I still need to acknowledge it as a game controller event. This is afaik how modern games like GTA5 handles it.

Here is a single button press from my controller:
```
type:SDL_CONTROLLERAXISMOTION axis:SDL_CONTROLLER_AXIS_TRIGGERLEFT value:20559
type:SDL_CONTROLLERAXISMOTION axis:SDL_CONTROLLER_AXIS_TRIGGERLEFT value:21587
type:SDL_CONTROLLERAXISMOTION axis:SDL_CONTROLLER_AXIS_TRIGGERLEFT value:22230
type:SDL_CONTROLLERAXISMOTION axis:SDL_CONTROLLER_AXIS_TRIGGERLEFT value:22487
type:SDL_CONTROLLERAXISMOTION axis:SDL_CONTROLLER_AXIS_TRIGGERLEFT value:22230
type:SDL_CONTROLLERAXISMOTION axis:SDL_CONTROLLER_AXIS_TRIGGERLEFT value:21202
type:SDL_CONTROLLERAXISMOTION axis:SDL_CONTROLLER_AXIS_TRIGGERLEFT value:19660
type:SDL_CONTROLLERAXISMOTION axis:SDL_CONTROLLER_AXIS_TRIGGERLEFT value:17475
type:SDL_CONTROLLERAXISMOTION axis:SDL_CONTROLLER_AXIS_TRIGGERLEFT value:15291
type:SDL_CONTROLLERAXISMOTION axis:SDL_CONTROLLER_AXIS_TRIGGERLEFT value:12592
type:SDL_CONTROLLERAXISMOTION axis:SDL_CONTROLLER_AXIS_TRIGGERLEFT value:5910
type:SDL_CONTROLLERAXISMOTION axis:SDL_CONTROLLER_AXIS_TRIGGERLEFT value:0
```
Previously that would result in 10 down AXIS_TRIGGERLEFT and 1 up AXIS_TRIGGERLEFT.

Reply:
Previously useBeltPotion would only accept an event every 300ms, but if you hold down the button for longer it would still result in multiple potions consumed for one click. Conversely, if you pressed it twice in 300ms (can happen when trying to heal a lot with small potions) only one of the clicks would register.

Most of this change is visable in https://github.com/diasurgical/devilutionX/pull/395/commits/1aa1a67ea712f004f57513ed688d7065f22486ce

Reply:
Makes sense, thank you